### ISSUE TITLE
Serial port failed to receive a byte

### ISSUE THREAD

**Issue by Alpha330i** [jaromir-sukuba/a-p-prog#16]
im trying to flash a PIC18F25k80 and always get stuck at the point, when serial port is failing to receive a byte. i connected all correct but im doing something wrong with software maybe?

C:\Users\...\sw>pp3.exe -c COM4 -s 1700 -t 18f25k80 text.hex
PP programmer, version 0.99
Sleeping for 1700 ms while arduino bootloader expires
Device ID: 6180
Programming FLASH (32768 B in 512 pages per 64 bytes):
Serial port failed to receive a byte, read returned 0
COM4: No error

**Comment by jaromir-sukuba** [comment]
Well, that is weird. Are you sure you have the newest firmware?

**Comment by Alpha330i** [comment]
Yes im trying with the latest firmware. Maybe its because the hex file is bigger than 32kb which Arduino is capable.
Hex File is 87kb

**Comment by InputBlackBoxOutput** [comment]
I got the same error!

I am using it to program PIC18F2550.  When I got the error, I used the -v 4 to get verbose output. The output is as follows: 
```text
PP programmer, version 0.99
Opening serial port
Opening filename blink.hex
File open

Read 44 chars: :100000000EEF00F00000000000EF00F00000000024
Line len 16 B, type 0, address 0x0000 offset 0x0000, EFF 0x000000
0EEF00F00000000000EF00F000000000


<More reads>

Read 12 chars: :00000001FF
Line len 0 B, type 1, address 0x0000 offset 0x0030, EFF 0x300000

Entering programming mode
TX: 0x01
TX: 0x00
RX: fail
Serial port failed to receive a byte, read returned 0
COM3: No error
```

The .hex file has a file size of 1kb

What we can deduce is that the file size is not causing the error. The error occurred when it entered the programming mode.

**Comment by WydD** [comment]
Heyo, I had the same issue using a Arduino Uno and a pretty fast computer.
I figured out that if your computer sends data too fast, and because the Uno doesnt have a lot of RAM (2k) and it cant handle the buffers size.

So, I worked around the issue by adding a `Sleep(1);` in `putByte`, it may be slower but it's still a few seconds to program a 16f1455.

Thanks for this project!

**Comment by Alpha330i** [comment]
> Heyo, I had the same issue using a Arduino Uno and a pretty fast computer.
> I figured out that if your computer sends data too fast, and because the Uno doesnt have a lot of RAM (2k) and it cant handle the buffers size.
> 
> So, I worked around the issue by adding a `Sleep(1);` in `putByte`, it may be slower but it's still a few seconds to program a 16f1455.
> 
> Thanks for this project!

Hello Thank you, in which line did you paste the sleep(1)?

**Comment by ffpp2003** [comment]
I figured out that after the progam fails one time, the second time always works. Could it be that the problem is that there is no enough time for the serial connection to get set up, and thats why always it fails to recive the byte on the first time, and not the subsequent ones?

**Comment by delight-f** [comment]
Can confirm I am getting this error too on a 16f1455.

**Comment by jaromir-sukuba** [comment]
I've seen the problem (Serial port failed to receive a byte, read returned 0) on a specific computer, not on other computers I ran this software.
As a workaround I found that opening the serial port with another program - in this case serial terminal - closing and then retrying pp3 software resulted in expected operation. I'm not sure yet what's the reason behind it - probably serial port initialization or buffers flushing is not sufficient in some cases and that causes the serial port to receive garbage, on the specific computer/OS. I'll investigate this some more.

**Comment by cjacker** [comment]
> Hi @jaromir-sukuba it seems that I have similar issue.
> 
> `./pp3 -c /dev/ttyACM0 -v 8 -t 16f1823 file.hex`
> 
> gives me:
> 
> ```
> Opening serial port
> Opening: /dev/ttyACM0 at 4097
> Opening filename file.hex 
> Entering programming mode
> TX: 0x01
> TX: 0x00
> RX: fail
> Serial port failed to receive a byte, read returned 0
> /dev/ttyACM0: No such file or directory
> ```
> 
> I also tried using -s 1000 which is a bit better:
> 
> ```
> Opening serial port
> Opening: /dev/ttyACM0 at 4097
> Sleeping for 1000 ms while arduino bootloader expires
> Opening filename file.hex 
> Entering programming mode
> TX: 0x01
> TX: 0x00
> RX:  0xF9
> getting ID for family 0
> Resetting PC
> TX: 0x03
> TX: 0x00
> RX: fail
> Serial port failed to receive a byte, read returned 0
> /dev/ttyACM0: No such file or directory
> ```
> 
> I added `sleep(1);` in getByte just before `read()` as suggested above, but unfortunately no luck.
> 
> I also tried disconnecting all cables from Arduino so it's not connected with PIC anymore and I'm getting exactly same results/outputs from pp3
> 
> my setup is:
> 
>     * Arduino UNO (official)
> 
>     * ArchLinux
> 
>     * PC wth AMD Ryzen 7 4800H
> 
>     * chip: PIC16F1823-E/P
> 
> 
> any ideas?

try increase sleep time like '-s 2000'?

I tested with PIC16F1823 and a China-made nano, it works with '-s 2000' without any issue.

**Comment by ready4droid** [comment]
I'm getting this error on my Arduino UNO as well, didn't have any issues with my chinese clone nano, but my uno gives me the "Serial port failed to receive a byte, read returned 0".  Oddly, it does Verifying FLASH, which looks to complete, and then gives me this error after it gets to Verifying config.
